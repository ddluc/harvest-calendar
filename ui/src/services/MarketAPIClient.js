import axios from 'axios';

class MarketAPIClient {

  constructor() {
    this.baseUrl = 'http://localhost:3300';
  }

  async getMarketData () {
    const url = `${this.baseUrl}/api/market_data`;
    try {
      const response = await axios.get(url);
      return response.data; 
    } catch (err) {
      console.error(err);
      return null;
    }
  }

}

export default new MarketAPIClient();
