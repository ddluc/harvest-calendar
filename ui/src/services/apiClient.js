/**
 * API Client
 */

import axios from 'axios';

const getBaseURL = () => {
  let baseURL = '';
  if (document.location.hostname === 'localhost') {
    baseURL = 'http://localhost:3300';
  } else {
    baseURL = 'http://harvestapp.ddluc.com';
  }
  return baseURL;
}

export const getMarketData = async () => {
  const url = `${getBaseURL()}/api/market_data`;
  try {
    const response = await axios.get(url);
    return response.data;
  } catch (err) {
    console.error(err);
    return null;
  }
}
