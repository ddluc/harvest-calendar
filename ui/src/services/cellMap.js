/**
 * Maps the market data to the grid cells
 */

export const buildCellMap = (activeSeason) => {

  let months = [];
  switch (activeSeason) {
    case 1:
      months = ['MAR', 'APR', 'MAY', 'JUN'];
      break;
    case 2:
      months = ['JUN', 'JUL', 'AUG', 'SEP'];
      break;
    case 3:
      months = ['SEP', 'OCT', 'NOV', 'DEC'];
      break;
    case 4:
      months = ['DEC', 'JAN', 'FEB', 'MAR'];
      break;
    default:
      months = ['MAR', 'APR', 'MAY', 'JUN'];
  }

  let cellMap = {body: {}, header: {}};
  let bodyCellIndex = 1;
  let headerCellIndex = 1;
  let segments = buildSegmentMap();
  months.forEach((month) => {
    cellMap.header[month] = headerCellIndex;
    headerCellIndex++;
    segments.forEach((segment) => {
      let key = `${month}-${segment.id}`;
      cellMap.body[key] = bodyCellIndex;
      bodyCellIndex++;
    });
   });
  return cellMap;

}

export const buildSegmentMap = () => {
  const segments = [
    {id: 1, key: 'early', label: 'Early'},
    {id: 2, key: 'mid', label: 'Mid', },
    {id: 3, key: 'late', label: 'Late'}
  ];
  return segments;
};
