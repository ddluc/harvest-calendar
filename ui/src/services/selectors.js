/**
 * Utility Methods
 **/

/**
 * getMonthsInSeason
 * @description gets the months for the active season
 * @param <Object> state - application state
 */
export const getMonthsInSeason = (state) => {
  let { months } = state.marketData;
  let { activeSeason } = state;
  let monthsInSeason = [];
  let order = [];
  switch (activeSeason) {
    // Spring
    case 1:
      order = ['MAR', 'APR', 'MAY', 'JUN'];
      break;
    // Summer
    case 2:
      order = ['JUN', 'JUL', 'AUG', 'SEP'];
      break;
    // Fall
    case 3:
      order = ['SEP', 'OCT', 'NOV', 'DEC'];
      break;
    // Winter
    case 4:
      order = ['DEC', 'JAN', 'FEB', 'MAR'];
      break;
    default:
      order = ['MAR', 'APR', 'MAY', 'JUN'];
  }
  order.forEach((key) => {
    months.forEach((month) => {
      if (month.key === key) {
        monthsInSeason.push(month);
      }
    });
  });
  return monthsInSeason;
};


/**
 * sortVegetables
 * @description sorts the vegetables by custom algorithm to determine how in season it is
 * @param <Object> state - application state
 */
export const sortVegetables = (state) => {
  let {marketData, activeMonth, activeSegment} = state;
  let peak = [];
  let early = [];
  let late = [];
  let off = [];
  let none = [];
  if (activeSegment) {
    marketData.vegetables.forEach((vegetable) => {
      vegetable.mPos.forEach((mPos) => {
        if (`${mPos.month}-${mPos.segment}` === `${activeMonth}-${activeSegment}` ) {
          if ( mPos.status === 'peak') peak.push(vegetable);
          if ( mPos.status === 'early') early.push(vegetable);
          if ( mPos.status === 'late') late.push(vegetable);
          if ( mPos.status === 'off') off.push(vegetable);
          if ( mPos.status === 'none') none.push(vegetable);
        }
      });
    });
    let sortedVegetables = [...peak, ...early, ...late, ...off, ...none ];
    return sortedVegetables;
  } else {
    return marketData.vegetables;
  }
};

/**
 * filterVegetables
 * @description filters the vegetables to the user's search query
 * @param <Object> state - application state
 */
export const filterVegetables = (state, vegetables) => {
  let { searchInput } = state;
  console.log(searchInput);
  console.log(vegetables);
  if (searchInput !== '' && searchInput.length > 2) {
    let filteredVegetables = [];
    vegetables.forEach((veg) => {
      if (veg.label.includes(searchInput)) {
        filteredVegetables.push(veg);
      }
    });
    return filteredVegetables;
  } else {
    return vegetables;
  }
}

/**
 * buildCellMap
 * @description maps the market data to the css grid
 * @param <Object> state - application state
 */
export const buildCellMap = (state) => {
  const segments = buildSegmentMap();
  const months = getMonthsInSeason(state);
  let cellMap = {body: {}, header: {}};
  let bodyCellIndex = 1;
  let headerCellIndex = 1;
  months.forEach((month) => {
    cellMap.header[month.key] = headerCellIndex;
    headerCellIndex++;
    segments.forEach((segment) => {
      let key = `${month.key}-${segment.id}`;
      cellMap.body[key] = bodyCellIndex;
      bodyCellIndex++;
    });
   });
  return cellMap;
}

export const buildSegmentMap = () => {
  const segments = [
    {id: 1, key: 'early', label: 'Early'},
    {id: 2, key: 'mid', label: 'Mid', },
    {id: 3, key: 'late', label: 'Late'}
  ];
  return segments;
};
