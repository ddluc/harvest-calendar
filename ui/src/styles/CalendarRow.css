.calendar-row {
  display: grid;
  grid-gap: 1px;
  height: 50px;
  grid-template-columns: 33% repeat(12, 1fr);
}

.calendar-cell {
  text-align: center;
}

.calendar-cell.cell-1 {
  text-align: left;
}
