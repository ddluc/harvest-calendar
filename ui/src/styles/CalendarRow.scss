@import '__theme.scss';

.calendar-row {
  display: grid;
  grid-template-columns: 26% repeat(12, 1fr);
  grid-template-rows: 50px;
}

.calendar-row:hover {
  border-top: inset 2px $charcol;
  border-bottom: inset 2px $charcol;
}

.calendar-cell.cell-label {
  order: -1;
  text-align: left;
  padding-left: 20px;
  color: $charcol;
  font-size: 16px;
  background-color: $white;
}

.calendar-cell.cell-1 {
  grid-row-start: 1;
  grid-row-end: 2;
  grid-column-start: 2;
  grid-column-end: 3;
}

.calendar-cell.cell-2 {
  grid-row-start: 1;
  grid-row-end: 2;
  grid-column-start: 3;
  grid-column-end: 4;
}

.calendar-cell.cell-3 {
  grid-row-start: 1;
  grid-row-end: 2;
  grid-column-start: 4;
  grid-column-end: 5;
}

.calendar-cell.cell-4 {
  grid-row-start: 1;
  grid-row-end: 2;
  grid-column-start: 5;
  grid-column-end: 6;
}

.calendar-cell.cell-5 {
  grid-row-start: 1;
  grid-row-end: 2;
  grid-column-start: 6;
  grid-column-end: 7;
}

.calendar-cell.cell-6 {
  grid-row-start: 1;
  grid-row-end: 2;
  grid-column-start: 7;
  grid-column-end: 8;
}

.calendar-cell.cell-7 {
  grid-row-start: 1;
  grid-row-end: 2;
  grid-column-start: 8;
  grid-column-end: 9;
}

.calendar-cell.cell-8 {
  grid-row-start: 1;
  grid-row-end: 2;
  grid-column-start: 9;
  grid-column-end: 10;
}

.calendar-cell.cell-9 {
  grid-row-start: 1;
  grid-row-end: 2;
  grid-column-start: 10;
  grid-column-end: 11;
}

.calendar-cell.cell-10 {
  grid-row-start: 1;
  grid-row-end: 2;
  grid-column-start: 11;
  grid-column-end: 12;
}

.calendar-cell.cell-11 {
  grid-row-start: 1;
  grid-row-end: 2;
  grid-column-start: 12;
  grid-column-end: 13;
}

.calendar-cell.cell-12 {
  grid-row-start: 1;
  grid-row-end: 2;
  grid-column-start: 13;
  grid-column-end: 14;
}
