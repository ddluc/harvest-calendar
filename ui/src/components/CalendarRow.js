import React from 'react';
import Cell from './Cell';
import { getMonthsInSeason } from '../services/UtilityService';

import '../styles/CalendarRow.css'

function CalendarRow({vegetable, marketData, view, activeSeason, activeMonth, cellMap, pricingMode}) {

  const months = getMonthsInSeason(marketData.months, activeSeason);

  const renderCells = () => {
    let cells = [];
    vegetable.mPos.forEach((mPos) => {
      let cellKey = `${mPos.month}-${mPos.segment}`;
      if (Object.keys(cellMap.body).includes(cellKey)) {
        let id = `${vegetable.key}-${cellKey}`;
        let cellIndex = cellMap.body[cellKey];
        cells.push(<Cell mPos={mPos} id={id} cellIndex={cellIndex} />);
      }
    });
    return cells;
  }

  return (
    <div className="calendar-row">
      <div className="calendar-cell cell-label">{vegetable.label}</div>
      { renderCells() }
    </div>
  );
}

export default CalendarRow;
