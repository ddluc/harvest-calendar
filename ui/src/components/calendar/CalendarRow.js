import React from 'react';
import Cell from './Cell';
import { Store } from '../../flux/store';
import { getMonthsInSeason } from '../../services/selectors';

import '../../styles/CalendarRow.scss'

function CalendarRow({vegetable, cellMap}) {
  const { state, dispatch } = React.useContext(Store);
  const months = getMonthsInSeason(state);
  let cells = [];
  vegetable.mPos.forEach((mPos) => {
    let cellKey = `${mPos.month}-${mPos.segment}`;
    if (Object.keys(cellMap.body).includes(cellKey)) {
      let id = `${vegetable.key}-${cellKey}`;
      let cellIndex = cellMap.body[cellKey];
      let isActive = (cellKey === `${state.activeMonth}-${state.activeSegment}`) ? 'active' : '';
      let showPrice = state.pricingMode === true ? 'showPrice' : '';
      cells.push(<Cell key={id} mPos={mPos} id={id} cellIndex={cellIndex} isActive={isActive} showPrice={showPrice}/>);
    }
  });

  return (
    <div className="calendar-row">
      <div className="calendar-cell cell-label">{vegetable.label}</div>
      { cells }
    </div>
  );
}

export default CalendarRow;
