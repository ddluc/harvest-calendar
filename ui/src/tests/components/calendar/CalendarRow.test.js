import React from 'react';
import CalendarRow from '../../../components/calendar/CalendarRow';
import * as selectors from '../../../services/selectors';
import { getMockState } from '../../__mocks/state.mock.js';
import {
  mockBuildSegmentMap,
  mockGetMonthsInSeason,
  mockBuildCellMap
} from '../../__mocks/selectors.mock.js';
import { shallow } from 'enzyme';

describe('CalendarRow', () => {

  it('should render', () => {
    let mockState = getMockState();
    // Random index vegetable array
    let vegetableIndex = 10;
    buildContextSpy(mockState);
    const props = {
      key: 1,
      months: mockGetMonthsInSeason(),
      vegetable: mockState.marketData.vegetables[vegetableIndex],
      cellMap: mockBuildCellMap(),
    };

    let wrapper = shallow(<CalendarRow {...props} />);
    expect(wrapper).toMatchSnapshot();
  })
});
