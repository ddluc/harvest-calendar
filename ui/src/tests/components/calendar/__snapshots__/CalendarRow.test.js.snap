// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarRow should render 1`] = `
<div
  className="calendar-row"
>
  <div
    className="calendar-cell cell-label"
  >
    <p>
      Celery
    </p>
    <small>
       
      Other
       | 
      <em>
        Apium graveolens
      </em>
    </small>
  </div>
  <Cell
    category="other"
    cellIndex={1}
    id="celery-MAR-1"
    isActive=""
    key="celery-MAR-1"
    mPos={
      Object {
        "id": 357,
        "month": "MAR",
        "segment": 1,
        "status": "none",
        "unit": "",
        "value": "0.00",
      }
    }
  />
  <Cell
    category="other"
    cellIndex={2}
    id="celery-MAR-2"
    isActive=""
    key="celery-MAR-2"
    mPos={
      Object {
        "id": 358,
        "month": "MAR",
        "segment": 2,
        "status": "none",
        "unit": "",
        "value": "0.00",
      }
    }
  />
  <Cell
    category="other"
    cellIndex={3}
    id="celery-MAR-3"
    isActive=""
    key="celery-MAR-3"
    mPos={
      Object {
        "id": 359,
        "month": "MAR",
        "segment": 3,
        "status": "none",
        "unit": "",
        "value": "0.00",
      }
    }
  />
  <Cell
    category="other"
    cellIndex={4}
    id="celery-APR-1"
    isActive=""
    key="celery-APR-1"
    mPos={
      Object {
        "id": 360,
        "month": "APR",
        "segment": 1,
        "status": "early",
        "unit": "bunch",
        "value": "1.99",
      }
    }
  />
  <Cell
    category="other"
    cellIndex={5}
    id="celery-APR-2"
    isActive=""
    key="celery-APR-2"
    mPos={
      Object {
        "id": 361,
        "month": "APR",
        "segment": 2,
        "status": "early",
        "unit": "bunch",
        "value": "1.99",
      }
    }
  />
  <Cell
    category="other"
    cellIndex={6}
    id="celery-APR-3"
    isActive=""
    key="celery-APR-3"
    mPos={
      Object {
        "id": 362,
        "month": "APR",
        "segment": 3,
        "status": "early",
        "unit": "bunch",
        "value": "1.99",
      }
    }
  />
  <Cell
    category="other"
    cellIndex={7}
    id="celery-MAY-1"
    isActive=""
    key="celery-MAY-1"
    mPos={
      Object {
        "id": 363,
        "month": "MAY",
        "segment": 1,
        "status": "early",
        "unit": "bunch",
        "value": "1.99",
      }
    }
  />
  <Cell
    category="other"
    cellIndex={8}
    id="celery-MAY-2"
    isActive="active"
    key="celery-MAY-2"
    mPos={
      Object {
        "id": 364,
        "month": "MAY",
        "segment": 2,
        "status": "peak",
        "unit": "bunch",
        "value": "1.99",
      }
    }
  />
  <Cell
    category="other"
    cellIndex={9}
    id="celery-MAY-3"
    isActive=""
    key="celery-MAY-3"
    mPos={
      Object {
        "id": 365,
        "month": "MAY",
        "segment": 3,
        "status": "peak",
        "unit": "bunch",
        "value": "1.99",
      }
    }
  />
  <Cell
    category="other"
    cellIndex={10}
    id="celery-JUN-1"
    isActive=""
    key="celery-JUN-1"
    mPos={
      Object {
        "id": 366,
        "month": "JUN",
        "segment": 1,
        "status": "peak",
        "unit": "bunch",
        "value": "1.99",
      }
    }
  />
  <Cell
    category="other"
    cellIndex={11}
    id="celery-JUN-2"
    isActive=""
    key="celery-JUN-2"
    mPos={
      Object {
        "id": 367,
        "month": "JUN",
        "segment": 2,
        "status": "peak",
        "unit": "bunch",
        "value": "1.99",
      }
    }
  />
  <Cell
    category="other"
    cellIndex={12}
    id="celery-JUN-3"
    isActive=""
    key="celery-JUN-3"
    mPos={
      Object {
        "id": 368,
        "month": "JUN",
        "segment": 3,
        "status": "peak",
        "unit": "bunch",
        "value": "1.99",
      }
    }
  />
</div>
`;
